planetary_computer/_pc_fs.py

```diff
@@ -0,0 +1,90 @@
+
+    Parameters
+    ----------
+    target_protocol : str
+        The protocol used to load the actual asset (e.g. 'https')
+    target_options : dict
+        Additional keywords to use for the target protocol's fsspec filesystem
+    fo: str, optional
+        The target path.
+
+    Examples
+    --------
+    This example loads a Kerchunk index file from Azure Blob Storage. The index file is in a private blob
+    storage container and so needs to be signed. The ``pc`` in the URL will automatically sign the
+    asset before attempting to load it.
+    
+    >>> import xarray as xr
+    >>> url = "reference::pc::https://deltaresreservoirssa.blob.core.windows.net/references/reservoirs/CHIRPS.json"
+    >>> result = xr.open_dataset(url, engine="zarr", consolidated=False)
+    <xarray.Dataset>
+    Dimensions:      (time: 13515, GrandID: 2951, ksathorfrac: 5)
+    Coordinates:
+      * GrandID      (GrandID) float64 nan nan nan nan nan ... nan nan nan nan nan
+      * ksathorfrac  (ksathorfrac) float64 5.0 20.0 50.0 100.0 250.0
+      * time         (time) datetime64[ns] NaT NaT NaT NaT NaT ... NaT NaT NaT NaT
+    Data variables: (12/14)
+        ETa          (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        Ea_res       (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        FracFull     (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        Melt         (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        P            (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        PET          (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        ...           ...
+        Qout_res     (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        S_res        (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        Snow         (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        Temp         (time, GrandID, ksathorfrac) float32 dask.array<chunksize=(1, 2951, 5), meta=np.ndarray>
+        latitude     (GrandID) float32 dask.array<chunksize=(2951,), meta=np.ndarray>
+        longitude    (GrandID) float32 dask.array<chunksize=(2951,), meta=np.ndarray>
+    """
+    def __init__(
+        self,
+        target_protocol=None,
+        target_options=None,
+        fo=None,
+        **kwargs,
+    ):
+        self.target_protocol = target_protocol
+        self.target_options = target_options
+        if fo:
+            fo = planetary_computer.sign(fo)
+        self.fo = fo
+        self.target_fs = fsspec.filesystem(self.target_protocol, **self.target_options)
+        if isinstance(self.target_fs, fsspec.implementations.reference.ReferenceFileSystem):
```

Comment:
I'm not a fan of this block.

The reference filesystem has the idea of "template" URLs, which are the NetCDF files in blob storage. We want to sign those URLs before anyone attempts to access data via this reference filesystem.

It seems that the reference filesystem's `__init__` calls a method at https://github.com/fsspec/filesystem_spec/blob/7effb83e8ab31010ec5796c14193b5fcd5774e05/fsspec/implementations/reference.py#L149, which does a lot of work including in-lining the `template` URLs in the reference `(url, start, end)` tuples. Unfortunately, we don't have a way to update the template URLs before the tuples are built, so we have to do it again.